src/config.cpp

```diff
@@ -478,5 +478,130 @@ const std::string& CircuitConfig::getExpandedJSON() const {
 }
 
 
+class SimulationConfig::Parser
+{
+  public:
+    Parser(const std::string& content, const std::string& basePath)
+        : _basePath(fs::absolute(fs::path(basePath)).lexically_normal())
+        , _json(nlohmann::json::parse(content)) {}
+
+    template <typename Iterator, typename Type, typename SectionName>
+    void parseMandatory(const Iterator& it,
+                        const char* name,
+                        const SectionName& sn,
+                        Type& buf) const {
+        const auto element = it.find(name);
+        if (element == it.end())
+            throw SonataError(fmt::format("Could not find '{}' in '{}'", name, sn));
+        buf = element->template get<Type>();
+    }
+
+    template <typename Iterator, typename Type>
+    void parseOptional(const Iterator& it, const char* name, Type& buf) const {
+        const auto element = it.find(name);
+        if (element != it.end())
+            buf = element->template get<Type>();
+    }
+
+    SimulationConfig::Run parseRun() const {
+        const auto runIt = _json.find("run");
+        if (runIt == _json.end())
+            throw SonataError("Could not find 'run' section");
+
+        SimulationConfig::Run result{};
+        parseMandatory(*runIt, "tstop", "run", result.tstop);
+        parseMandatory(*runIt, "dt", "run", result.dt);
+        return result;
+    }
+
+    SimulationConfig::Output parseOutput() const {
+        SimulationConfig::Output result{};
+        result.outputDir = "output";
+        result.spikesFile = "out.h5";
+
+        const auto outputIt = _json.find("output");
+        if (outputIt != _json.end()) {
+            parseOptional(*outputIt, "output_dir", result.outputDir);
+            parseOptional(*outputIt, "spikes_file", result.spikesFile);
+        }
+
+        result.outputDir = toAbsolute(_basePath, result.outputDir);
+
+        return result;
+    }
+
+    std::unordered_map<std::string, SimulationConfig::Report> parseReports() const {
+        std::unordered_map<std::string, SimulationConfig::Report> result;
+
+        const auto reportsIt = _json.find("reports");
+        if (reportsIt == _json.end())
+            return result;
+
+        for (auto it = reportsIt->begin(); it != reportsIt->end(); ++it) {
+            auto& report = result[it.key()];
+            auto& valueIt = it.value();
+            const auto debugStr = fmt::format("report {}", it.key());
+            parseMandatory(valueIt, "cells", debugStr, report.cells);
+            parseMandatory(valueIt, "type", debugStr, report.type);
+            parseMandatory(valueIt, "dt", debugStr, report.dt);
+            parseMandatory(valueIt, "start_time", debugStr, report.startTime);
+            parseMandatory(valueIt, "end_time", debugStr, report.endTime);
```

Comment:
The problem in this task is to know exactly what to do when we are missing data, what what defaults to apply. See e.g. `/gpfs/bbp.cscs.ch/project/proj12/example_data/sonata/v2/usecase3/simulation_sonata.json`. This is a minimalistic config file, in principle valid, and won't pass here.

Reply:
I think we can resolve this conversation: mandatory fields need to be present in the files.